### ISSUE TITLE
[PAID] [$1000] If you grab an image and leave it outside image area then the image will remain in the grabbed state and also show console error - reported by @Tushu17

### ISSUE THREAD

**Issue by mvtglobally** [Expensify/App#11192]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Open Website.
2. Go to chat > send a jpeg/png attachment > open that attachment.
3. Zoom & grab the image and leave the grab outside of the image.
4. If you close the preview box, you can see console error.

## Expected Result:
Cursor shouldn't remain in grabbed state even if user leave it outside image area.

## Actual Result:
It remains in the grabbed state and also show console error.

## Workaround:
unknown 

## Platform:

Where is this issue occurring?

- Web
- Desktop App

**Version Number:** 1.2.1-0
**Reproducible in staging?:** Y 
**Reproducible in production?:** Y 
**Email or phone of affected tester (no customers):** 
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation 

https://user-images.githubusercontent.com/43995119/191658891-3f511d43-336e-4754-b19b-1bebdfd53e92.mov

**Expensify/Expensify Issue URL:** 
**Issue reported by:** @Tushu17
**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1661962009275619

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by JmillsExpensify** [comment]
Issue looks good. I'm going to apply the engineering label and keep myself assigned.

**Comment by NikkiWines** [comment]
Definitely looks like a good one for upwork! Adding `External` label

**Comment by roryabraham** [comment]
Fun!

**Comment by Tushu17** [comment]
###Proposal

We need to use `onMouseOut` prop of `Pressable` 


https://github.com/Expensify/App/blob/2bef1705d5aac60c9463cc0f080155dc9925d033/src/components/ImageView/index.js#L247-L250

````diff
    changeGrabState() {
        if (!this.state.isDragging) {
            return;
        }

        this.setState({isDragging: false, isMouseDown: false});
    }
   ````
   
   and we need to pass `changeGrabState` in onMouseOut

**Comment by mananjadhav** [comment]
@Tushu17 While your proposal works, problem is, if I drag into the image, after dragging out and just leave the mouse it'll zoom out the image.

**Comment by Tushu17** [comment]
@mananjadhav, The solution doesn't touch the `isZoomed` state which is responsible for image zoom in or zoom out. So It's working fine.

https://user-images.githubusercontent.com/89839952/192859800-4f09691e-a0c5-47db-81e3-3e565556ac78.mov

**Comment by mananjadhav** [comment]
@Tushu17 Are you leaving the cursor anytime outside the image and coming back?

**Comment by Tushu17** [comment]
@mananjadhav Yes, In the video I was leaving cursor outside image to show that it doesn't zoom out.

> if I drag into the image, after dragging out and just leave the mouse it'll zoom out the image.

Can you share a screen recording of this?

**Comment by mananjadhav** [comment]
@Tushu17 here's the video for you after applying your proposal.

https://user-images.githubusercontent.com/3069065/193448577-6da69ca3-7a8c-4da8-9bd4-d286e21572ec.mp4

**Comment by Tushu17** [comment]
@mananjadhav , I couldn't understand the problem you are trying to convey through the video. Can you explain the bug and It's repro streps?

**Comment by mananjadhav** [comment]
After applying your proposal:

1. Open any attachment in the chat
2. Click on the + button to zoom in the image
3. Press the mouse and drag the image around.
4. Without lifting the mouse, take the cursor outside the image view to confirm image doesn't drag (your code fixes this)
5. Without lifting the mouse, take the cursor inside the image view 
*6. Lift the pressed mouse key, the image zooms out (this shouldn't happen)*

Hope this clarifies @Tushu17

**Comment by JmillsExpensify** [comment]
Still working through the ideal proposal.

**Comment by JmillsExpensify** [comment]
@Tushu17 will you be updating your proposal based on @mananjadhav's feedback?

**Comment by Tushu17** [comment]
### Updated Proposal

We are not capturing events that are happening outside container thats why when user takes cursor outside image in  `isMouseDown` as `true` and un-click mouse there, it doesn't not change `isMouseDown` state to `false`.
### Solution 1
We can solve this problem by using `e.nativeEvent.buttons`, if it turns 0 we can change `isMouseDown` state to `false`.
https://github.com/Expensify/App/blob/ad33dd75603f6d19346f4fa38fc996557ab78551/src/components/ImageView/index.js#L176
```
if (e.nativeEvent.buttons === 0 && this.state.isMouseDown) {
            this.setState({isMouseDown: false, isDragging: false});
        }
```
### Solution 2

Now if we consider dragging image from outside as a problem then we can use two props of `Pressable`  `onMouseOut` & `onMouseEnter` and a state `isMouseClicked` to check if user has changed his mouse state outside image or not.

 So the final solution will be to add :-
1. We need to add
````
if (e.nativeEvent.buttons === 0) {
            this.setState({isMouseClicked: false});
        }
  ````
in `trackMovement `
https://github.com/Expensify/App/blob/ad33dd75603f6d19346f4fa38fc996557ab78551/src/components/ImageView/index.js#L176

2.
`onMouseOut` will prevent user from dragging image from Outside by changing `isMouseDown` to `false` and will also change the `isMouseClicked` state to `true` if user has moved out of image while dragging.

`onMouseEnter` will use `isMouseClicked` state to know if user has un-clicked his mouse outside image area. If user hasn't un-clicked his mouse or we can say `isMouseClicked` is 'true' then if user come over image, we can change `isMouseDown` state to `true` to keep cursor in grabbed form.
 ````
changeGrabState() {
        if (!this.state.isDragging) {
            return;
        }
        this.setState({isDragging: false, isMouseDown: false, isMouseClicked: true});
    }
    
    mouseEnterImage(e) {
        if (e.nativeEvent.buttons === 0 || !this.state.isMouseClicked) {
            return;
        }
        this.setState({isMouseDown: true, isDragging: true});
    }
````
and we need to pass `changeGrabState` to `onMouseOut` and `mouseEnterImage` to `onMouseEnter`

**Comment by JmillsExpensify** [comment]
@mananjadhav ready for another review of the proposal?

**Comment by mananjadhav** [comment]
@Tushu17 Thanks for the update proposal. Are you suggestion Solution 1 and Solution 2 as two distinct proposals? or they're part 1 and 2 of the whole solution?

Because just applying Solution 1 didn't work for me.

Solution 2 does work but it does have a problem (I can't reproduce this on staging and I am not sure if its a browser behavior). Attached is the video. I've tried my best to mention the reproduction steps (hope it helps)


https://user-images.githubusercontent.com/3069065/194484276-e73500d8-235b-4ecf-a088-4b7be449e410.mp4


1. Open any attachment in the chat
2. Click on the + button to zoom in the image
3. Keep the attachment in the zoom out state and mouse your move in an around the image.
4. Without lifting the mouse, take the cursor in the image,
5. Lift the mouse and click again to zoom the image
6. Once the image is zoomed, drag the mouse and you'll see Image doesn't get dragged and rather the scroll of the outer containers are dragged, inferring the drag is performed on the container than the image.

**Comment by Tushu17** [comment]
> Are you suggestion Solution 1 and Solution 2 as two distinct proposals? or they're part 1 and 2 of the whole solution?

Solution 1 and Solution 2 are different proposal as they both give different results. Solution 1 won't stop user from dragging image from outside of container. It will just change `isMouseDown` state to `false` if user leaves mouse click. 

video after applying solution 1 :-

https://user-images.githubusercontent.com/89839952/194497058-1c716ed9-1264-4930-9bfe-fce61d76e80a.mov

Imo solution 1 is good as it give user more space to drag image but incase you believe that dragging image from outside isn't expected then we can use Solution 2 after which user won't be able to drag image from outside.
 
 Video after solution 2 :- 
 
https://user-images.githubusercontent.com/89839952/194498478-b8c37dd2-950c-40af-bc55-d14820318aaa.mov

> Solution 2 does work but it does have a problem (I can't reproduce this on staging and I am not sure if its a browser behavior). Attached is the video. I've tried my best to mention the reproduction steps (hope it helps)

After reading repro steps, I tried my best to repro this but couldn't succeed. Is it very frequently reproducible with you?

**Comment by mananjadhav** [comment]
Yes @Tushu17 I am able to reproduce this frequently. You can check the attached video. 

And about Solution 1, I guess our core requirement is that we don't want the image to be in the grabbed state.

@JmillsExpensify Should we double? We haven't received an appropriate proposal so far ?

**Comment by JmillsExpensify** [comment]
Yes, agreed. We should double. I'll do that right now.

**Comment by JmillsExpensify** [comment]
Price update in this issue and in Upwork as well: https://www.upwork.com/jobs/~01941d2f457766f5e4.

**Comment by JmillsExpensify** [comment]
Still waiting on proposals

**Comment by JmillsExpensify** [comment]
Doubling the price again to hopefully get more proposals and interest.

**Comment by luispdev0328** [comment]
I could reproduce this issue when send an attachment, too.

https://user-images.githubusercontent.com/115640297/196430132-1a79f6ad-ba86-44cc-8b60-9abff9d2827f.mp4

Do you need my proposal yet?

**Comment by mananjadhav** [comment]
@luispdev0328 Yes its the same component so solving it once, will solve at both the places.  Yes is it open for proposals. Feel free to send one for review.

**Comment by mdneyazahmad** [comment]
## Proposal


https://github.com/Expensify/App/blob/c3a0129ee0b1e6fb524e65bf15361624d71d9a14/src/components/ImageView/index.js#L52-L60

`document.addEventListener('mousemove', this.trackMovement.bind(this)); ` the reference to listener is not preserved here so when `document.removeEventListener('mousemove', this.trackMovement.bind(this)); ` is called it does nothing because that is not the actual reference of the listener. Thus the listener is kept calling even if the modal is closed, and `scrollableRef` does not exist inside `trackMovement` thus the console error appears.

```diff
+        this.onMouseUp = this.onMouseUp.bind(this);

-        document.addEventListener('mousemove', this.trackMovement.bind(this));
+        document.addEventListener('mousemove', this.trackMovement);

-        document.removeEventListener('mousemove', this.trackMovement.bind(this));
+        document.removeEventListener('mousemove', this.trackMovement);
```

https://github.com/Expensify/App/blob/c3a0129ee0b1e6fb524e65bf15361624d71d9a14/src/components/ImageView/index.js#L105-L107

When mouse is button is released `onContainerPress` is called. There it checks whether to perform click behaviour or stop the drag. As `onContainerPress` is attached to `Pressable` we miss this event if the cursor is released outside this element. We will add one more event listener `mouseup` and test if the cursor is released outside the `Pressable`. If so we will reset the drag state.

```diff
+        this.onMouseUp = this.onMouseUp.bind(this);

+        document.addEventListener('mouseup', this.onMouseUp);

+        document.removeEventListener('mouseup', this.onMouseUp);

+    onMouseUp(e) {
+        const isInside = this.scrollableRef.contains(e.nativeEvent.target);
+
+        if (!isInside && this.state.isZoomed && this.state.isDragging && this.state.isMouseDown) {
+            this.setState({isDragging: false, isMouseDown: false});
+        }
+    }
```

<details>
<summary>Diff</summary>

```diff
diff --git a/src/components/ImageView/index.js b/src/components/ImageView/index.js
index 84cce4f29..744552bac 100644
--- a/src/components/ImageView/index.js
+++ b/src/components/ImageView/index.js
@@ -25,6 +25,8 @@ class ImageView extends PureComponent {
         this.onContainerPress = this.onContainerPress.bind(this);
         this.imageLoadingStart = this.imageLoadingStart.bind(this);
         this.imageLoadingEnd = this.imageLoadingEnd.bind(this);
+        this.onMouseUp = this.onMouseUp.bind(this);
+        this.trackMovement = this.trackMovement.bind(this);
         this.state = {
             isLoading: false,
             containerHeight: 0,
@@ -49,7 +51,8 @@ class ImageView extends PureComponent {
         if (this.canUseTouchScreen) {
             return;
         }
-        document.addEventListener('mousemove', this.trackMovement.bind(this));
+        document.addEventListener('mousemove', this.trackMovement);
+        document.addEventListener('mouseup', this.onMouseUp);
     }
 
     componentWillUnmount() {
@@ -57,7 +60,8 @@ class ImageView extends PureComponent {
             return;
         }
 
-        document.removeEventListener('mousemove', this.trackMovement.bind(this));
+        document.removeEventListener('mousemove', this.trackMovement);
+        document.removeEventListener('mouseup', this.onMouseUp);
     }
 
     /**
@@ -116,6 +120,14 @@ class ImageView extends PureComponent {
         }
     }
 
+    onMouseUp(e) {
+        const isInside = this.scrollableRef.contains(e.nativeEvent.target);
+
+        if (!isInside && this.state.isZoomed && this.state.isDragging && this.state.isMouseDown) {
+            this.setState({isDragging: false, isMouseDown: false});
+        }
+    }
+
     /**
      * When open image, set image width, height.
      * @param {Number} imageWidth

```
</details>

**Comment by mananjadhav** [comment]
Thanks for the proposal @mdneyazahmad. I like the idea of tracking mouseup and `scrollableRef.contains`. The only thing is I applied your fix and it didn't work for me. Did it work for you? Did we miss something?

**Comment by mdneyazahmad** [comment]
@mananjadhav it works for me. Let me know what issues are you getting.

commit: https://github.com/mdneyazahmad/App/commit/f93e41b5b2b17a8f4cac7fae44b4ea957f36f696
branch: https://github.com/mdneyazahmad/App/tree/fix/11192-drag-image

https://user-images.githubusercontent.com/77761491/197117592-36d26b98-d8b0-4f87-aace-b3d85978ab16.mp4

**Comment by JmillsExpensify** [comment]
@roryabraham thoughts on @mdneyazahmad's proposal above? Once we agree, I'll get this assigned and we can assign/hire in Upwork as well.

**Comment by roryabraham** [comment]
@mdneyazahmad looks good except for the naming the method `onMouseUp` – recall that member functions should be named for what they do, not what they handle.

**Comment by JmillsExpensify** [comment]
Thanks @mdneyazahmad! I've just hired you in Upwork. In addition, @mananjadhav please review PR and make sure to apply to the [Upwork job](https://www.upwork.com/jobs/~01941d2f457766f5e4) when you can.

**Comment by mananjadhav** [comment]
I've reviewed the changes, they look fine. I'll get done with the test by tomorrow.

**Comment by JmillsExpensify** [comment]
All active contributors hired. @Tushu17 I've created an [Upwork job](https://www.upwork.com/jobs/~0174a18f261d0f5b8f) here for the reporting bonus. Please apply!

**Comment by Tushu17** [comment]
@JmillsExpensify Okay applied, Thanks.

**Comment by JmillsExpensify** [comment]
Thanks, hired!

**Comment by JmillsExpensify** [comment]
@roryabraham I know sequence numbers is higher priority, though this PR is a good one to help move forward when you're able.

**Comment by roryabraham** [comment]
PR merged

**Comment by JmillsExpensify** [comment]
Then separately, @roryabraham @mananjadhav do ya'll mind helping confirm whether root cause analysis is needed? Based on that, we can decide if/what convos should be moved to contributor-plus. 

> @mananjadhav @roryabraham The PR that introduced the bug has been identified. Link to the PR:
> @mananjadhav @roryabraham The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
> @mananjadhav @roryabraham A discussion in #contributor-plus has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:

**Comment by roryabraham** [comment]
> whether we want to add a new regression test to TestRail

Yes, we should, using the reproduction steps from the OP of this issue.

> do ya'll mind helping confirm whether root cause analysis is needed?

I don't think so. It's not clear that this is a regression, it might have just been a bug that was always there.

**Comment by JmillsExpensify** [comment]
Let's circle back to the OP, because I'm a bit skeptical on the test we want to run. Basically:

1. Log into NewDot on web
2. Navigate to any chat
3. Click the grey plus button to the left of the chat input
4.  Choose add attachment and choose a jpeg/png file
5. Open the attachment
6. Zoom & grab the image, leaving the grab outside of the image
7. **Verify** that the cursor doesn't remain in a grabbed state

**Comment by JmillsExpensify** [comment]
All contributors paid out at this point. I'm still not sure if any regression testing would be helpful, so waiting for a response there because checking that last item off the list.

**Comment by mananjadhav** [comment]
Thanks @JmillsExpensify

**Comment by roryabraham** [comment]
> Let's circle back to the OP, because I'm a bit skeptical on the test we want to run. Basically:

Log into NewDot on web
Navigate to any chat
Click the grey plus button to the left of the chat input
Choose add attachment and choose a jpeg/png file
Open the attachment
Zoom & grab the image, leaving the grab outside of the image
Verify that the cursor doesn't remain in a grabbed state

These steps seem accurate, I don't see why we wouldn't want to add them to the regression suite.

**Comment by JmillsExpensify** [comment]
Disregard the PDFs test, as that's for another issue. Regression test has been added and my BZ items are checked off, so I'm closing this issue.
